File: ios/Classes/AppRestriction/MethodChannel/RestrictionsMethodHandler.swift
import Flutter
import Foundation

final class RestrictionsMethodHandler {
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case MethodNames.configureShield:
            handleConfigureShield(call: call, result: result)
        case MethodNames.setRestrictedApps:
            handleSetRestrictedApps(call: call, result: result)
        case MethodNames.addRestrictedApp:
            handleAddRestrictedApp(call: call, result: result)
        case MethodNames.removeRestriction:
            handleRemoveRestriction(call: call, result: result)
        case MethodNames.isRestricted:
            handleIsRestricted(call: call, result: result)
        case MethodNames.removeAllRestrictions:
            handleRemoveAllRestrictions(call: call, result: result)
        case MethodNames.getRestrictedApps:
            handleGetRestrictedApps(call: call, result: result)
        case MethodNames.isRestrictionSessionActiveNow:
            handleIsRestrictionSessionActiveNow(result: result)
        case MethodNames.isRestrictionSessionConfigured:
            handleIsRestrictionSessionConfigured(result: result)
        case MethodNames.pauseEnforcement:
            handlePauseEnforcement(call: call, result: result)
        case MethodNames.resumeEnforcement:
            handleResumeEnforcement(result: result)
        case MethodNames.getRestrictionSession:
            handleGetRestrictionSession(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleConfigureShield(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard var configuration = call.arguments as? [String: Any] else {
            result(PluginErrors.invalidArguments(PluginErrorMessage.missingShieldConfiguration))
            return
        }

        let appGroupId = configuration["appGroupId"] as? String
        AppGroupStore.updateGroupIdentifier(appGroupId)
        configuration.removeValue(forKey: "appGroupId")

        if let typedData = configuration["iconBytes"] as? FlutterStandardTypedData {
            configuration["iconBytes"] = typedData.data
        } else if configuration["iconBytes"] is NSNull {
            configuration.removeValue(forKey: "iconBytes")
        }

        switch ShieldConfigurationStore.storeConfiguration(configuration, appGroupId: appGroupId) {
        case .success:
            result(nil)
        case .appGroupUnavailable(let resolvedGroupId):
            var details: [String: Any] = [
                "resolvedAppGroupId": resolvedGroupId
            ]
            if let appGroupId {
                details["appGroupId"] = appGroupId
            } else {
                details["appGroupId"] = NSNull()
            }
            result(PluginErrors.appGroupError(details: details))
        }
    }

    private func handleSetRestrictedApps(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(PluginErrorMessage.restrictionsUnsupported))
            return
        }
        guard let args = call.arguments as? [String: Any],
              let tokens = args["identifiers"] as? [String] else {
            result(PluginErrors.invalidArguments(PluginErrorMessage.missingIdentifiers))
            return
        }
        let decodeResult = ShieldManager.shared.decodeTokens(base64Tokens: tokens)
        if !decodeResult.invalidTokens.isEmpty {
            result(PluginErrors.invalidToken(
                message: PluginErrorMessage.unableToDecodeTokens,
                invalidTokens: decodeResult.invalidTokens
            ))
            return
        }

        switch RestrictionStateStore.storeDesiredRestrictedApps(decodeResult.appliedBase64Tokens) {
        case .success:
            break
        case .appGroupUnavailable(let resolvedGroupId):
            result(appGroupError(resolvedGroupId: resolvedGroupId))
            return
        }

        applyDesiredRestrictionsIfNeeded()
        result(decodeResult.appliedBase64Tokens)
    }

    private func handleAddRestrictedApp(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(PluginErrorMessage.restrictionsUnsupported))
            return
        }
        guard let args = call.arguments as? [String: Any],
              let token = args["identifier"] as? String else {
            result(PluginErrors.invalidArguments(PluginErrorMessage.missingIdentifier))
            return
        }

        let decodeResult = ShieldManager.shared.decodeTokens(base64Tokens: [token])
        if !decodeResult.invalidTokens.isEmpty {
            result(PluginErrors.invalidToken(
                message: PluginErrorMessage.unableToDecodeToken,
                invalidTokens: decodeResult.invalidTokens
            ))
            return
        }

        ensureDesiredRestrictionsInitializedFromManagedStore()
        var desired = RestrictionStateStore.loadDesiredRestrictedApps()
        if desired.contains(token) {
            result(false)
            return
        }
        desired.append(token)

        switch RestrictionStateStore.storeDesiredRestrictedApps(desired) {
        case .success:
            break
        case .appGroupUnavailable(let resolvedGroupId):
            result(appGroupError(resolvedGroupId: resolvedGroupId))
            return
        }

        applyDesiredRestrictionsIfNeeded()
        result(true)
    }

    private func handleRemoveRestriction(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(PluginErrorMessage.restrictionsUnsupported))
            return
        }
        guard let args = call.arguments as? [String: Any],
              let token = args["identifier"] as? String else {
            result(PluginErrors.invalidArguments(PluginErrorMessage.missingIdentifier))
            return
        }

        let decodeResult = ShieldManager.shared.decodeTokens(base64Tokens: [token])
        if !decodeResult.invalidTokens.isEmpty {
            result(PluginErrors.invalidToken(
                message: PluginErrorMessage.unableToDecodeToken,
                invalidTokens: decodeResult.invalidTokens
            ))
            return
        }

        ensureDesiredRestrictionsInitializedFromManagedStore()
        var desired = RestrictionStateStore.loadDesiredRestrictedApps()
        let previousCount = desired.count
        desired.removeAll { $0 == token }
        let changed = desired.count != previousCount

        switch RestrictionStateStore.storeDesiredRestrictedApps(desired) {
        case .success:
            break
        case .appGroupUnavailable(let resolvedGroupId):
            result(appGroupError(resolvedGroupId: resolvedGroupId))
            return
        }

        applyDesiredRestrictionsIfNeeded()
        result(changed)
    }

    private func handleIsRestricted(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(PluginErrorMessage.restrictionsUnsupported))
            return
        }
        guard let args = call.arguments as? [String: Any],
              let token = args["identifier"] as? String else {
            result(PluginErrors.invalidArguments(PluginErrorMessage.missingIdentifier))
            return
        }

        let decodeResult = ShieldManager.shared.decodeTokens(base64Tokens: [token])
        if !decodeResult.invalidTokens.isEmpty {
            result(PluginErrors.invalidToken(
                message: PluginErrorMessage.unableToDecodeToken,
                invalidTokens: decodeResult.invalidTokens
            ))
            return
        }
        ensureDesiredRestrictionsInitializedFromManagedStore()
        let restricted = RestrictionStateStore.loadDesiredRestrictedApps().contains(token)
        result(restricted)
    }

    private func handleRemoveAllRestrictions(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(PluginErrorMessage.restrictionsUnsupported))
            return
        }

        switch RestrictionStateStore.storeDesiredRestrictedApps([]) {
        case .success:
            break
        case .appGroupUnavailable(let resolvedGroupId):
            result(appGroupError(resolvedGroupId: resolvedGroupId))
            return
        }

        ShieldManager.shared.clearRestrictions()
        result(nil)
    }

    private func handleGetRestrictedApps(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result([])
            return
        }
        ensureDesiredRestrictionsInitializedFromManagedStore()
        let tokens = RestrictionStateStore.loadDesiredRestrictedApps()
        result(tokens)
    }

    private func handleIsRestrictionSessionActiveNow(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(false)
            return
        }

        ensureDesiredRestrictionsInitializedFromManagedStore()
        applyDesiredRestrictionsIfNeeded()
        let restrictedApps = RestrictionStateStore.loadDesiredRestrictedApps()
        let isPausedNow = RestrictionStateStore.loadPausedUntilEpochMs() > 0
        result(!restrictedApps.isEmpty && !isPausedNow)
    }

    private func handleIsRestrictionSessionConfigured(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(false)
            return
        }

        ensureDesiredRestrictionsInitializedFromManagedStore()
        let restrictedApps = RestrictionStateStore.loadDesiredRestrictedApps()
        result(!restrictedApps.isEmpty)
    }

    private func handlePauseEnforcement(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(PluginErrorMessage.restrictionsUnsupported))
            return
        }

        guard let args = call.arguments as? [String: Any],
              let durationValue = args["durationMs"] as? NSNumber else {
            result(PluginErrors.invalidArguments("Missing or invalid 'durationMs' argument"))
            return
        }
        let durationMs = durationValue.int64Value
        if durationMs <= 0 {
            result(PluginErrors.invalidArguments("Missing or invalid 'durationMs' argument"))
            return
        }

        if RestrictionStateStore.loadPausedUntilEpochMs() > 0 {
            result(PluginErrors.invalidArguments("Restriction enforcement is already paused"))
            return
        }

        let pausedUntilEpochMs = RestrictionStateStore.currentEpochMs() + durationMs
        switch RestrictionStateStore.storePausedUntilEpochMs(pausedUntilEpochMs) {
        case .success:
            break
        case .appGroupUnavailable(let resolvedGroupId):
            result(appGroupError(resolvedGroupId: resolvedGroupId))
            return
        }

        ShieldManager.shared.clearRestrictions()
        result(nil)
    }

    private func handleResumeEnforcement(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(PluginErrorMessage.restrictionsUnsupported))
            return
        }

        switch RestrictionStateStore.storePausedUntilEpochMs(0) {
        case .success:
            break
        case .appGroupUnavailable(let resolvedGroupId):
            result(appGroupError(resolvedGroupId: resolvedGroupId))
            return
        }

        applyDesiredRestrictionsIfNeeded()
        result(nil)
    }

    private func handleGetRestrictionSession(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result([
                "isActiveNow": false,
                "isPausedNow": false,
                "pausedUntilEpochMs": NSNull(),
                "restrictedApps": [String]()
            ])
            return
        }

        ensureDesiredRestrictionsInitializedFromManagedStore()
        applyDesiredRestrictionsIfNeeded()
        let restrictedApps = RestrictionStateStore.loadDesiredRestrictedApps()
        let pausedUntilEpochMs = RestrictionStateStore.loadPausedUntilEpochMs()
        let isPausedNow = pausedUntilEpochMs > 0
        result([
            "isActiveNow": !restrictedApps.isEmpty && !isPausedNow,
            "isPausedNow": isPausedNow,
            "pausedUntilEpochMs": isPausedNow ? pausedUntilEpochMs : NSNull(),
            "restrictedApps": restrictedApps
        ])
    }

    @available(iOS 16.0, *)
    private func applyDesiredRestrictionsIfNeeded() {
        ensureDesiredRestrictionsInitializedFromManagedStore()
        let desiredRestrictedApps = RestrictionStateStore.loadDesiredRestrictedApps()
        if desiredRestrictedApps.isEmpty {
            ShieldManager.shared.clearRestrictions()
            return
        }

        let isPausedNow = RestrictionStateStore.loadPausedUntilEpochMs() > 0
        if isPausedNow {
            ShieldManager.shared.clearRestrictions()
            return
        }

        let decodeResult = ShieldManager.shared.decodeTokens(base64Tokens: desiredRestrictedApps)
        if !decodeResult.invalidTokens.isEmpty {
            ShieldManager.shared.clearRestrictions()
            return
        }
        ShieldManager.shared.setRestrictedApps(decodeResult.tokens)
    }

    private func appGroupError(resolvedGroupId: String) -> FlutterError {
        PluginErrors.appGroupError(
            details: [
                "resolvedAppGroupId": resolvedGroupId,
                "appGroupId": AppGroupStore.currentGroupIdentifier
            ]
        )
    }

    @available(iOS 16.0, *)
    private func ensureDesiredRestrictionsInitializedFromManagedStore() {
        let desiredRestrictedApps = RestrictionStateStore.loadDesiredRestrictedApps()
        if !desiredRestrictedApps.isEmpty {
            return
        }

        let currentlyApplied = ShieldManager.shared.getRestrictedApps()
        if currentlyApplied.isEmpty {
            return
        }

        _ = RestrictionStateStore.storeDesiredRestrictedApps(currentlyApplied)
    }
}

File: android/src/main/kotlin/com/example/pauza_screen_time/app_restriction/method_channel/RestrictionsMethodHandler.kt
package com.example.pauza_screen_time.app_restriction.method_channel

import android.content.Context
import com.example.pauza_screen_time.app_restriction.RestrictionManager
import com.example.pauza_screen_time.app_restriction.ShieldOverlayManager
import com.example.pauza_screen_time.core.MethodNames
import com.example.pauza_screen_time.core.PluginErrorHelper
import com.example.pauza_screen_time.permissions.PermissionHandler
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class RestrictionsMethodHandler(
    private val contextProvider: () -> Context?
) : MethodCallHandler {
    companion object {
        private const val ANDROID_ACCESSIBILITY_KEY = "android.accessibility"
        private const val FEATURE = "restrictions"
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                MethodNames.CONFIGURE_SHIELD -> handleConfigureShield(call, result)
                MethodNames.SET_RESTRICTED_APPS -> handleSetRestrictedApps(call, result)
                MethodNames.ADD_RESTRICTED_APP -> handleAddRestrictedApp(call, result)
                MethodNames.REMOVE_RESTRICTION -> handleRemoveRestriction(call, result)
                MethodNames.REMOVE_ALL_RESTRICTIONS -> handleRemoveAllRestrictions(result)
                MethodNames.GET_RESTRICTED_APPS -> handleGetRestrictedApps(result)
                MethodNames.IS_RESTRICTED -> handleIsRestricted(call, result)
                MethodNames.IS_RESTRICTION_SESSION_ACTIVE_NOW -> handleIsRestrictionSessionActiveNow(result)
                MethodNames.IS_RESTRICTION_SESSION_CONFIGURED -> handleIsRestrictionSessionConfigured(result)
                MethodNames.PAUSE_ENFORCEMENT -> handlePauseEnforcement(call, result)
                MethodNames.RESUME_ENFORCEMENT -> handleResumeEnforcement(result)
                MethodNames.GET_RESTRICTION_SESSION -> handleGetRestrictionSession(result)
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = call.method,
                message = "Unexpected restriction error: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleConfigureShield(call: MethodCall, result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.CONFIGURE_SHIELD,
                message = "Application context is not available",
            )
            return
        }

        val configMap = call.arguments as? Map<String, Any?>
        if (configMap == null) {
            PluginErrorHelper.invalidArgument(
                result = result,
                feature = FEATURE,
                action = MethodNames.CONFIGURE_SHIELD,
                message = "Shield configuration map is required",
            )
            return
        }

        try {
            ShieldOverlayManager.getInstance(context).configure(configMap)
            result.success(null)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.CONFIGURE_SHIELD,
                message = "Failed to configure shield: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleSetRestrictedApps(call: MethodCall, result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.SET_RESTRICTED_APPS,
                message = "Application context is not available",
            )
            return
        }

        val args = call.arguments
        val identifiers: List<String>? = when (args) {
            is Map<*, *> -> {
                val raw = args["identifiers"]
                if (raw is List<*>) {
                    val list = raw.filterIsInstance<String>()
                    if (list.size == raw.size) list else null
                } else {
                    null
                }
            }
            is List<*> -> {
                val list = args.filterIsInstance<String>()
                if (list.size == args.size) list else null
            }
            else -> null
        }

        if (identifiers == null) {
            PluginErrorHelper.invalidArgument(
                result = result,
                feature = FEATURE,
                action = MethodNames.SET_RESTRICTED_APPS,
                message = "Missing or invalid 'identifiers' argument",
            )
            return
        }

        try {
            val trimmed = identifiers.map { it.trim() }
            val hasBlank = trimmed.any { it.isBlank() }
            if (hasBlank) {
                PluginErrorHelper.invalidArgument(
                    result = result,
                    feature = FEATURE,
                    action = MethodNames.SET_RESTRICTED_APPS,
                    message = "Identifiers must be non-blank strings",
                )
                return
            }

            val applied = LinkedHashSet<String>()
            for (identifier in trimmed) {
                applied.add(identifier)
            }
            val appliedList = applied.toList()

            if (appliedList.isNotEmpty()) {
                val missingPrerequisites = getMissingPrerequisites(context)
                if (missingPrerequisites.isNotEmpty()) {
                    PluginErrorHelper.missingPermission(
                        result = result,
                        feature = FEATURE,
                        action = MethodNames.SET_RESTRICTED_APPS,
                        message = "Restriction prerequisites are not satisfied",
                        missing = missingPrerequisites,
                        status = mapOf(
                            ANDROID_ACCESSIBILITY_KEY to PermissionHandler(context)
                                .checkPermission(PermissionHandler.ACCESSIBILITY_KEY),
                        ),
                    )
                    return
                }
            }

            RestrictionManager.getInstance(context).setRestrictedApps(appliedList)
            result.success(appliedList)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.SET_RESTRICTED_APPS,
                message = "Failed to set restricted apps: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleAddRestrictedApp(call: MethodCall, result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.ADD_RESTRICTED_APP,
                message = "Application context is not available",
            )
            return
        }

        val identifier = call.argument<String>("identifier")
        if (identifier == null) {
            PluginErrorHelper.invalidArgument(
                result = result,
                feature = FEATURE,
                action = MethodNames.ADD_RESTRICTED_APP,
                message = "Missing or invalid 'identifier' argument",
            )
            return
        }

        try {
            val missingPrerequisites = getMissingPrerequisites(context)
            if (missingPrerequisites.isNotEmpty()) {
                PluginErrorHelper.missingPermission(
                    result = result,
                    feature = FEATURE,
                    action = MethodNames.ADD_RESTRICTED_APP,
                    message = "Restriction prerequisites are not satisfied",
                    missing = missingPrerequisites,
                    status = mapOf(
                        ANDROID_ACCESSIBILITY_KEY to PermissionHandler(context)
                            .checkPermission(PermissionHandler.ACCESSIBILITY_KEY),
                    ),
                )
                return
            }

            val added = RestrictionManager.getInstance(context).addRestrictedApp(identifier)
            result.success(added)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.ADD_RESTRICTED_APP,
                message = "Failed to add restricted app: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleRemoveRestriction(call: MethodCall, result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.REMOVE_RESTRICTION,
                message = "Application context is not available",
            )
            return
        }

        val identifier = call.argument<String>("identifier")
        if (identifier == null) {
            PluginErrorHelper.invalidArgument(
                result = result,
                feature = FEATURE,
                action = MethodNames.REMOVE_RESTRICTION,
                message = "Missing or invalid 'identifier' argument",
            )
            return
        }

        try {
            val removed = RestrictionManager.getInstance(context).removeRestriction(identifier)
            result.success(removed)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.REMOVE_RESTRICTION,
                message = "Failed to remove restriction: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleRemoveAllRestrictions(result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.REMOVE_ALL_RESTRICTIONS,
                message = "Application context is not available",
            )
            return
        }

        try {
            RestrictionManager.getInstance(context).removeAllRestrictions()
            result.success(null)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.REMOVE_ALL_RESTRICTIONS,
                message = "Failed to remove all restrictions: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleGetRestrictedApps(result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.GET_RESTRICTED_APPS,
                message = "Application context is not available",
            )
            return
        }

        try {
            val apps = RestrictionManager.getInstance(context).getRestrictedApps()
            result.success(apps)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.GET_RESTRICTED_APPS,
                message = "Failed to get restricted apps: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleIsRestricted(call: MethodCall, result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.IS_RESTRICTED,
                message = "Application context is not available",
            )
            return
        }

        val identifier = call.argument<String>("identifier")
        if (identifier == null) {
            PluginErrorHelper.invalidArgument(
                result = result,
                feature = FEATURE,
                action = MethodNames.IS_RESTRICTED,
                message = "Missing or invalid 'identifier' argument",
            )
            return
        }

        try {
            val isRestricted = RestrictionManager.getInstance(context).isRestricted(identifier)
            result.success(isRestricted)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.IS_RESTRICTED,
                message = "Failed to check if app is restricted: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleIsRestrictionSessionActiveNow(result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.IS_RESTRICTION_SESSION_ACTIVE_NOW,
                message = "Application context is not available",
            )
            return
        }

        try {
            val restrictionManager = RestrictionManager.getInstance(context)
            val restrictedApps = restrictionManager.getRestrictedApps()
            val isPausedNow = restrictionManager.isPausedNow()
            val isPrerequisitesMet = areRestrictionPrerequisitesMet(context)
            result.success(restrictedApps.isNotEmpty() && !isPausedNow && isPrerequisitesMet)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.IS_RESTRICTION_SESSION_ACTIVE_NOW,
                message = "Failed to get restriction session active state: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleIsRestrictionSessionConfigured(result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.IS_RESTRICTION_SESSION_CONFIGURED,
                message = "Application context is not available",
            )
            return
        }

        try {
            val restrictedApps = RestrictionManager.getInstance(context).getRestrictedApps()
            result.success(restrictedApps.isNotEmpty())
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.IS_RESTRICTION_SESSION_CONFIGURED,
                message = "Failed to get restriction session configuration state: ${e.message}",
                error = e,
            )
        }
    }

    private fun handlePauseEnforcement(call: MethodCall, result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.PAUSE_ENFORCEMENT,
                message = "Application context is not available",
            )
            return
        }

        val durationMs = call.argument<Number>("durationMs")?.toLong()
        if (durationMs == null || durationMs <= 0L) {
            PluginErrorHelper.invalidArgument(
                result = result,
                feature = FEATURE,
                action = MethodNames.PAUSE_ENFORCEMENT,
                message = "Missing or invalid 'durationMs' argument",
            )
            return
        }

        try {
            val restrictionManager = RestrictionManager.getInstance(context)
            if (restrictionManager.isPausedNow()) {
                PluginErrorHelper.invalidArgument(
                    result = result,
                    feature = FEATURE,
                    action = MethodNames.PAUSE_ENFORCEMENT,
                    message = "Restriction enforcement is already paused",
                )
                return
            }

            restrictionManager.pauseFor(durationMs)
            ShieldOverlayManager.getInstanceOrNull()?.hideShield()
            result.success(null)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.PAUSE_ENFORCEMENT,
                message = "Failed to pause restriction enforcement: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleResumeEnforcement(result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.RESUME_ENFORCEMENT,
                message = "Application context is not available",
            )
            return
        }

        try {
            RestrictionManager.getInstance(context).clearPause()
            result.success(null)
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.RESUME_ENFORCEMENT,
                message = "Failed to resume restriction enforcement: ${e.message}",
                error = e,
            )
        }
    }

    private fun handleGetRestrictionSession(result: Result) {
        val context = contextProvider()
        if (context == null) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.GET_RESTRICTION_SESSION,
                message = "Application context is not available",
            )
            return
        }

        try {
            val restrictionManager = RestrictionManager.getInstance(context)
            val restrictedApps = restrictionManager.getRestrictedApps()
            val pausedUntilEpochMs = restrictionManager.getPausedUntilEpochMs()
            val isPausedNow = pausedUntilEpochMs > 0L
            val isPrerequisitesMet = areRestrictionPrerequisitesMet(context)
            result.success(
                mapOf(
                    "isActiveNow" to (restrictedApps.isNotEmpty() && !isPausedNow && isPrerequisitesMet),
                    "isPausedNow" to isPausedNow,
                    "pausedUntilEpochMs" to if (isPausedNow) pausedUntilEpochMs else null,
                    "restrictedApps" to restrictedApps,
                )
            )
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = MethodNames.GET_RESTRICTION_SESSION,
                message = "Failed to get restriction session: ${e.message}",
                error = e,
            )
        }
    }

    private fun areRestrictionPrerequisitesMet(context: Context): Boolean {
        return getMissingPrerequisites(context).isEmpty()
    }

    private fun getMissingPrerequisites(context: Context): List<String> {
        val permissionHandler = PermissionHandler(context)
        val accessibilityStatus = permissionHandler.checkPermission(PermissionHandler.ACCESSIBILITY_KEY)
        if (accessibilityStatus == PermissionHandler.STATUS_GRANTED) {
            return emptyList()
        }
        return listOf(ANDROID_ACCESSIBILITY_KEY)
    }

}
